jupyter_client/provisioning.py

```diff
@@ -0,0 +1,407 @@
+from .localinterfaces import is_local_ip, local_ips
+
+DEFAULT_PROVISIONER = "LocalProvisioner"
+
+
+class KernelProvisionerMeta(ABCMeta, type(LoggingConfigurable)):
+    pass
+
+
+class KernelProvisionerBase(ABC, LoggingConfigurable, metaclass=KernelProvisionerMeta):
+    """Base class defining methods for KernelProvisioner classes.
+
+       Theses methods model those of the Subprocess Popen class:
+       https://docs.python.org/3/library/subprocess.html#popen-objects
+    """
+    # The kernel specification associated with this provisioner
+    kernel_spec: Any
+    kernel_id: str
+
+    def __init__(self, kernel_id: str, kernel_spec: Any, **kwargs):
+        self.kernel_id = kernel_id
+        self.kernel_spec = kernel_spec
+        super().__init__(**kwargs)
+
+    @abstractmethod
+    async def poll(self) -> [int, None]:
+        """Checks if kernel process is still running.
+
+         If running, None is returned, otherwise the process's integer-valued exit code is returned.
+         """
+        pass
+
+    @abstractmethod
+    async def wait(self) -> [int, None]:
+        """Waits for kernel process to terminate."""
+        pass
+
+    @abstractmethod
+    async def send_signal(self, signum: int) -> None:
+        """Sends signal identified by signum to the kernel process."""
+        pass
+
+    @abstractmethod
+    async def kill(self, restart=False) -> None:
+        """Kills the kernel process.  This is typically accomplished via a SIGKILL signal, which
+        cannot be caught.
+
+        restart is True if this operation precedes a start launch_kernel request.
+        """
+        pass
+
+    @abstractmethod
+    async def terminate(self, restart=False) -> None:
+        """Terminates the kernel process.  This is typically accomplished via a SIGTERM signal, which
+        can be caught, allowing the kernel provisioner to perform possible cleanup of resources.
+
+        restart is True if this operation precedes a start launch_kernel request.
+        """
+        pass
+
+    @abstractmethod
+    async def launch_kernel(self, cmd: List[str], **kwargs: Any) -> Tuple['KernelProvisionerBase', Dict]:
+        """Launch the kernel process returning the class instance and connection info."""
+        pass
+
+    async def cleanup(self, restart=False) -> None:
+        """Cleanup any resources allocated on behalf of the kernel provisioner.
+
+        restart is True if this operation precedes a start launch_kernel request.
+        """
+        pass
+
+    async def pre_launch(self, **kwargs: Any) -> Dict[str, Any]:
+        """Perform any steps in preparation for kernel process launch.
+
+        This includes applying additional substitutions to the kernel launch command and env.
+        It also includes preparation of launch parameters.
+
+        Returns potentially updated kwargs.
+        """
+
+        env = kwargs.pop('env', os.environ).copy()
+        # Don't allow PYTHONEXECUTABLE to be passed to kernel process.
```

Comment:
Probably should generalize this in the future for other kernels (not a priority)

Reply:
Are you referring to _an_ ability for provisioners to scrub the environment passed to the kernel?  That is a good idea, although the provisioner won't necessarily know what kernel it's provisioning.  I suspect the provisioner could use the kernel_spec's `language` value (if present) to make some inference though.   

Perhaps a method to the effect of `_finalize_env()` might be helpful to subclassed provisioners?

Reply:
Yeah if we're doing a major version bump it might be nice to have the abstraction expanded so it's at least partially flexible to non ipython kernel contexts here. To avoid scope creep I'd keep it in a separate PR though if we do tackle it